libyara/modules/math.c

```diff
@@ -0,0 +1,496 @@
+      uint64_t data_len = min(length, block->size - data_offset);
+      total_len += data_len;
+
+      offset += data_len;
+      length -= data_len;
+
+      for (i = 0; i < data_len; i++)
+      {
+        uint8_t c = *(block->data + data_offset + i);
+        data[c] += 1;
+      }
+
+      past_first_block = true;
+    }
+    else if (past_first_block)
+    {
+      // If offset is not within current block and we already
+      // past the first block then the we are trying to compute
+      // the checksum over a range of non contiguos blocks. As
+      // range contains gaps of undefined data the checksum is
+      // undefined.
+
+      yr_free(data);
+      return_double(UNDEFINED);
+    }
+
+    if (block->base + block->size > offset + length)
+      break;
+  }
+
+  if (!past_first_block)
+  {
+    yr_free(data);
+    return_double(UNDEFINED);
+  }
+
+  for (i = 0; i < 256; i++)
+  {
+    if (data[i] != 0)
+    {
+      x = (double) (data[i]) / total_len;
+      entropy -= x * log2(x);
+    }
+  }
+  yr_free(data);
+  return_double(entropy);
+}
+
+
+define_function(string_arithmetic_mean)
+{
+  int i;
+  double sum = 0.0;
+  SIZED_STRING* s = sized_string_argument(1);
+
+  if (IS_UNDEFINED(s))
+    return_double(UNDEFINED);
+
+  for (i = 0; i < s->length; i++)
+    sum += (double) s->c_string[i];
+
+
+  sum = sum / (double) s->length;
+  return_double(fabs((sum - 127.5) / 127.5));
+}
+
+
+define_function(data_arithmetic_mean)
+{
+  int i;
+  double sum = 0.0;
+  bool past_first_block = false;
+  uint64_t total_len = 0;
+
+  int64_t offset = integer_argument(1);
+  int64_t length = integer_argument(2);
+
+  if (IS_UNDEFINED(offset) || IS_UNDEFINED(length))
+    return_double(UNDEFINED);
+
+  YR_SCAN_CONTEXT* context = scan_context();
+  YR_MEMORY_BLOCK* block = NULL;
+
+  if (offset < 0 || length < 0 || offset < context->mem_block->base)
+  {
+    return ERROR_WRONG_ARGUMENTS;
+  }
+
+  foreach_memory_block(context, block)
+  {
+    if (offset >= block->base &&
+        offset < block->base + block->size)
+    {
+      uint64_t data_offset = offset - block->base;
+      uint64_t data_len = min(length, block->size - data_offset);
+      total_len += data_len;
+
+      offset += data_len;
+      length -= data_len;
+
+      for (i = 0; i < data_len; i++)
+        sum += (double) *(block->data + data_offset + i);
+
+      past_first_block = true;
+    }
+    else if (past_first_block)
+    {
+      // If offset is not within current block and we already
+      // past the first block then the we are trying to compute
+      // the checksum over a range of non contiguos blocks. As
+      // range contains gaps of undefined data the checksum is
+      // undefined.
+      return_double(UNDEFINED);
+    }
+
+    if (block->base + block->size > offset + length)
+      break;
+  }
+
+  if (!past_first_block)
+    return_double(UNDEFINED);
+
+  sum = sum / (double) total_len;
+  return_double(fabs((sum - 127.5) / 127.5));
```

Comment:
I find this function confusing, because it's called arithmetic_mean but it actually returns a deviation percentage from the mean. What I propose is creating a function named "deviation" receiving the string (or offset, lenght pair) and a number between 0 and 255 which will be the central point. The function will return the deviation defined by:

![image](https://cloud.githubusercontent.com/assets/182937/5595086/95e197a0-9269-11e4-8509-2895b186de15.png)

Where m(X) is our central point. Use the mean as m(X) and you'll have the standard deviation.

This will give you what you want in a more standard way that people familiar with statistical language will recognise. I would also rename "arithmetic_mean" to just "mean", to keep names as shorter as possible.